## Problem Set 1.2 

In [87]:
import numpy as np
import math

In [29]:
def unit_vector(vector):
    """ Returns the unit vector of the vector.  """
    return vector / np.linalg.norm(vector)

def angle_between(v1, v2):
    """ Returns the angle in radians between vectors 'v1' and 'v2'::

            >>> angle_between((1, 0, 0), (0, 1, 0))
            1.5707963267948966
            >>> angle_between((1, 0, 0), (1, 0, 0))
            0.0
            >>> angle_between((1, 0, 0), (-1, 0, 0))
            3.141592653589793
    """
    v1_u = unit_vector(v1)
    v2_u = unit_vector(v2)
    return np.arccos(np.clip(np.dot(v1_u, v2_u), -1.0, 1.0))

In [54]:
def angle(v1, v2):
    # angle = 57.2968 * radian
    return (angle_between(v1, v2) * 57.2968)

In [68]:
def length(v):
    return np.linalg.norm(v)

### №1

In [2]:
# dot products u,v; u,w; u,v+w; w,v.
u = np.array([-.6, .8])
v = np.array([4, 3])
w = np.array([1, 2])

print(round(np.dot(u,v)))
print(round(np.dot(u,w)))
print(round(np.dot(u,v+w)))
print(round(np.dot(w,v)))

0.0
1.0
1.0
10


### №2 

In [3]:
# magnitudes u,v,w.
print(np.linalg.norm(u))
print(np.linalg.norm(v))
print(np.linalg.norm(w))

1.0
5.0
2.23606797749979


In [4]:
# dot product of u,v; product of magnitudes of u,v.
print(abs(round(np.dot(u,v))), np.linalg.norm(u) * np.linalg.norm(v))

0.0 5.0


In [5]:
# dot product of v,w; product of magnitudes of v,w.
print(abs(round(np.dot(v,w))), round(np.linalg.norm(v) * np.linalg.norm(w), 4))

10 11.1803


### №3

In [6]:
# unit vectors of v,w; angle between v,w.
v_unit = v/np.linalg.norm(v)
print(v_unit)

w_unit = w/np.linalg.norm(w)
print(w_unit)

radian_vw = np.arccos(np.dot(v_unit, w_unit))
angle_vw = radian_vw * 57.2958
print(round(angle_vw,2))

[0.8 0.6]
[0.4472136  0.89442719]
26.57


* *0°: w itself*
* *90°: any orthogonal vector to w*
* *180°: -w*

### №4

In [7]:
# dot products of any unit vectors v,w:
# 1. v, -v 
# 2. v+w, v-w
# 3. v-2w, v+2w

# example of unit vectors:
v = np.array([0.8, 0.6])
w = np.array([0.6, 0.8])

print(np.dot(v, -v))
print(np.dot(v+w, v-w))
print(np.dot(v-2*w, v+2*w))

-1.0
0.0
-3.0


* *v · (-v) = -1*
* *(v + w) · (v - w) = 0*
* *(v - 2w) · (v + 2w) = -3*

### №5

In [8]:
# find unit vectors of v,w; assign them to u1,u2.
v = np.array([1,3])
w = np.array([2,1,2])

u1 = v/np.linalg.norm(v)
u2 = w/np.linalg.norm(w)

print(np.around(u1, 4), np.around(u2, 4))

[0.3162 0.9487] [0.6667 0.3333 0.6667]


In [9]:
# find unit vectors U1 and U2 that are perpendicular to u1 and u2

#### *Let U1 components be x1, x2*  
*Solve equation: 0.3162·x1 + 0.9487·x2 = 0*  
*x1 = -3·x2*  
* *Let x2 be 1, then x1 = -3*  
*U1 = (-3, 1) / ||U1||*  
* *Let x2 be -1, then x1 = 3*  
*U1 = (3, -1) / ||U1||*  
  
  

#### *Let U2 components be y1, y2, y3*  
*Solve equation: 0.6667·y1 + 0.3333·y2 + 0.6667·y3 = 0*  
*The units vectors of the whole plain can be perpendicular to U2*

### №6

* *2·w1 + (-1)·w2 = 0*  
*2·w1 = w2*  
* *The vectors perpendicluar to (1,1,1) lies on a plane in R3*
* *The vectors perpendicular to (1,1,1) and (1,2,3) lies on a line in R3*

### №7

In [33]:
#a
v = np.array([1, math.sqrt(3)])
w = np.array([1, 0])

print(round(angle(v,w)))

60.0


In [32]:
#b
v = np.array([2, 2, -1])
w = np.array([2, -1, 2])

print(round(angle(v,w)))

90.0


In [34]:
#c
v = np.array([1, math.sqrt(3)])
w = np.array([-1, math.sqrt(3)])

print(round(angle(v,w)))

60.0


In [35]:
#d
v = np.array([3, 1])
w = np.array([-1, -2])

print(round(angle(v,w)))

135.0


### №8

* *If u = (1, 1, 1) is perpendicular to v and w, then vis parallel tow*  
*False*  
* *If u is perpendicular to v and w, then u is perpendicular to v + 2w*  
*True*  
* *If u and v are perpendicular unit vectors then ||u - v|| = sqrt(2)*  
*True*  

In [39]:
u = np.array([0,1])
v = np.array([1,0])

print(np.linalg.norm(u-v), math.sqrt(2))

1.4142135623730951 1.4142135623730951


### №9

*If v2·w2/v1·w1 = −1*  
*then v2·w2 = −v1·w1 or v1·w1+v2·w2 = v·w = 0: perpendicular!*

In [42]:
v1 = 1
v2 = 4*v1
v = np.array([v1,v2])

w1 = 1
w2 = -1/4*w1
w = np.array([w1,w2])

print(round(angle(v, w)))

print(v2*w2/v1*w1)

90.0
-1.0


### №10

In [48]:
v = np.array([1,2])
w = np.array([-2,1])

print(v[1]/v[0] * w[1]/w[0])
print(np.dot(v,w)/(np.linalg.norm(v)*np.linalg.norm(w)))
print(round(angle(v,w)))

-1.0
0.0
90.0


### №11

*v · w < 0 means angle > 90◦*  
*these w’s fill half of 3-dimensional space*

### №12

In [50]:
v = np.array([1,1])
w = np.array([1,5])

# c = v·w / v·v
c = np.dot(v,w)/np.dot(v,v)
print(c)

3.0


### №13

*Find perpendicular vector to (1,0,1)*  
*Answer: all vectors (c, d, −c)*

### №14

*Find nonzero vectors u, v, w that are perpendicular to (1, 1, 1, 1) and to each other*

*Answer: u = (1, −1, 0, 0), v = (0, 0, 1, −1), w = (1, 1, −1, −1)
and (1, 1, 1, 1) are perpendicular to each other*

### №15

In [53]:
u = np.array([math.sqrt(2),math.sqrt(8)])
w = np.array([math.sqrt(8),math.sqrt(2)])

cosine = np.dot(u,w) / (np.linalg.norm(u) * np.linalg.norm(u))
print(cosine)

0.8


### №16

In [55]:
v = np.array([1,1,1,1,1,1,1,1,1])

magnitude = np.linalg.norm(v)
print(magnitude)

v_unit = v/magnitude
print(v_unit)

3.0
[0.33333333 0.33333333 0.33333333 0.33333333 0.33333333 0.33333333
 0.33333333 0.33333333 0.33333333]


### №18

In [76]:
v = np.array([4,2])
w = np.array([-1,2])

len_v = length(v)
len_w = length(w)
len_vw = length(v+w)

print(np.around([len_v**2, len_w**2, len_vw**2]))

[20.  5. 25.]


### №19

In [84]:
#example
v = np.array([5,4])
w = np.array([7,2])
u = np.array([4,4])
c = 10

# v·w = w·v
print("1.", np.dot(v,w), "=", np.dot(w,v))
# u·(v+w)= u·v + u·w
print("2.", np.dot(u,v+w), "=", np.dot(u,v), "+", np.dot(u,w))
# (cv)·w = c(v·w)
print("3.", np.dot(c*v,w), "=", c*np.dot(v,w))

1. 43 = 43
2. 72 = 36 + 36
3. 430 = 430


### №20

In [90]:
v = np.array([5,4])
w = np.array([7,2])

# ||v-w||^2
print(round(length(v-w)**2))

# ||v||^2 - 2||v||*||w||*cos_vw + ||w||^2
print(round(length(v)**2 - 2*length(v)*length(w)*np.cos(angle_between(v,w)) + length(w)**2))

8.0
8.0
